### PR TITLE
Vision docs 📝

### PR DESCRIPTION
**Pull request by merveenoyan**
This is a big PR to refresh outdated parts of vision docs and add new ones for newer tasks. 

Here's some action items:
- [x] add SAM2 FT to mask-generation tutorial
- [ ] add trackio to vision examples (requires extended support from Trainer)
- [ ] add universal segmentation docs (@ariG23498 is taking it up)
- [x] fix dataset in semantic segmentation doc (there was a canonical one which didn't load, fixed it)
- [x] rewrite video-text-to-text completely with video processors
- [x] update outdated image-text-to-text docs
- [ ] (maybe) swap models with newer ones

### CODE REVIEW

**Comment by molbap** [comment]
Subscribed, thanks for starting this! Feel free to ping me for a review

**Comment by merveenoyan** [comment]
@molbap @stevhliu I did a revamp of the docs for a few tasks, I might do follow-up but I don't want to be overwhelmed and rather keep this PR atomic, so feel free to review.

**Comment by merveenoyan** [comment]
@stevhliu would be great if you can review!

File: docs/source/en/tasks/image_text_to_text.md [hunk 1]
```diff
@@ -65,24 +66,29 @@ The image inputs look like the following.
...
 
-```python
-from PIL import Image
-import requests
 
-img_urls =["https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cats.png",
-           "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/bee.jpg"]
-images = [Image.open(requests.get(img_urls[0], stream=True).raw),
-          Image.open(requests.get(img_urls[1], stream=True).raw)]
+Below is an example of the chat template. For single prompt, we can create our chat template like below, providing image URLs directly with text prompt.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_text_to_text.md]: Clearer to not use "chat template" since this is just a chat history.

```suggestion
Structure your conversation as shown below for a single prompt with image and text inputs.
```

File: docs/source/en/tasks/image_text_to_text.md [hunk 2]
```diff
@@ -65,24 +66,29 @@ The image inputs look like the following.
...
+        "content": [
+            {"type": "image", "image": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cats.png"},
+            {"type": "text", "text": "What do we see in this image?"},
+        ]
+    }
+]
 ```
 
-Below is an example of the chat template. We can feed conversation turns and the last message as an input by appending it at the end of the template.
+In case we want to feed conversation turns to condition the generation better (like few-shot prompting), we can add assistant response ourselves, and the actual prompt as an input at the end of the template.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_text_to_text.md]:
```suggestion
Alternate between the `user` and `assistant` role to ground the model with prior context to generate better responses.
```

File: docs/source/en/tasks/image_text_to_text.md [hunk 3]
```diff
@@ -289,19 +296,43 @@ VLMs are often large and need to be optimized to fit on smaller hardware. Transf
 First, install dependencies.
 
 ```bash
-pip install -U quanto bitsandbytes
+pip install -U optimum-quanto bitsandbytes
 ```
 
-To quantize a model during loading, we need to first create [`QuantoConfig`]. Then load the model as usual, but pass `quantization_config` during model initialization.
+To quantize a model during loading, we need to first create [`QuantoConfig`]. Then load the model as usual, but pass `quantization_config` during model initialization. On top of this, you can compile a model using `torch.compile()` for additional speed gains.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_text_to_text.md]: I think its not recommended to use `torch.compile` outside of `generate` to avoid triggering recompilation.

File: docs/source/en/tasks/image_text_to_text.md [hunk 4]
```diff
@@ -289,19 +296,43 @@ VLMs are often large and need to be optimized to fit on smaller hardware. Transf
...
+            {"type": "image", "image": "https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/cats.png"},
+            {"type": "text", "text": "What do we see in this image?"},
+        ]
+    },
+]
+inputs = processor.apply_chat_template(messages, add_generation_prompt=True, tokenize=True, return_dict=True, return_tensors="pt").to(model.device)
+input_len = len(inputs.input_ids[0])
+
+with torch.no_grad():
+    generated_ids = model.generate(**inputs, max_new_tokens=100)
```
Comment (stevhliu) [inline docs/source/en/tasks/image_text_to_text.md]: This automatically triggers compilation

```suggestion
    generated_ids = model.generate(**inputs, cache_implementation="static", max_new_tokens=100)
```

File: docs/source/en/tasks/image_text_to_text.md [hunk 5]
```diff
@@ -289,19 +296,43 @@ VLMs are often large and need to be optimized to fit on smaller hardware. Transf
...
-model_id = "HuggingFaceM4/idefics2-8b"
+model_id = "Qwen/Qwen3-VL-4B-Instruct"
 quantization_config = QuantoConfig(weights="int8")
 quantized_model = AutoModelForImageTextToText.from_pretrained(
     model_id, device_map="auto", quantization_config=quantization_config
 )
+quantized_model = torch.compile(quantized_model)
+```
+
+Infer like usual.
```
Comment (stevhliu) [inline docs/source/en/tasks/image_text_to_text.md]:
```suggestion
```

File: docs/source/en/tasks/mask_generation.md [hunk 6]
```diff
@@ -24,8 +24,9 @@ Mask generation models are trained on large amounts of data and operate in two m
 - Prompting mode: In this mode, the model takes in an image and a prompt, where a prompt can be a 2D point location (XY coordinates) in the image within an object or a bounding box surrounding an object. In prompting mode, the model only returns the mask over the object
 that the prompt is pointing out.
 - Segment Everything mode: In segment everything, given an image, the model generates every mask in the image. To do so, a grid of points is generated and overlaid on the image for inference.
+- Video Inference: The model takes in a video, and a point or box prompt in a video frame, which is tracked throughout the video. You can get more information on how to do video inference by following [SAM 2 docs](model_doc/sam2).
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
- Video Inference: The model accepts a video, and a point or box prompt in a video frame, which is tracked throughout the video. You can get more information on how to do video inference by following [SAM 2 docs](../model_doc/sam2).
```

File: docs/source/en/tasks/mask_generation.md [hunk 7]
```diff
@@ -24,8 +24,9 @@ Mask generation models are trained on large amounts of data and operate in two m
...
 
-Mask generation task is supported by [Segment Anything Model (SAM)](model_doc/sam). It's a powerful model that consists of a Vision Transformer-based image encoder, a prompt encoder, and a two-way transformer mask decoder. Images and prompts are encoded, and the decoder takes these embeddings and generates valid masks.
+Mask generation task is supported by [Segment Anything Model (SAM)](model_doc/sam) and [Segment Anything Model 2 (SAM2)](model_doc/sam2), while video inference is supported by [Segment Anything Model 2 (SAM2)](model_doc/sam2). SAM is a powerful model that consists of a Vision Transformer-based image encoder, a prompt encoder, and a two-way transformer mask decoder. Images and prompts are encoded, and the decoder takes these embeddings and generates valid masks.  Meanwhile SAM 2 extends SAM by adding a memory module to track the masks.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
Mask generation task is supported by [Segment Anything Model (SAM)](../model_doc/sam) and [Segment Anything Model 2 (SAM2)](../model_doc/sam2), while video inference is supported by [Segment Anything Model 2 (SAM2)](../model_doc/sam2). SAM is a powerful model that consists of a Vision Transformer-based image encoder, a prompt encoder, and a two-way transformer mask decoder. Images and prompts are encoded, and the decoder takes these embeddings and generates valid masks.  Meanwhile, SAM 2 extends SAM by adding a memory module to track the masks. 
```

File: docs/source/en/tasks/mask_generation.md [hunk 8]
```diff
@@ -134,7 +127,7 @@ processor = SamProcessor.from_pretrained("facebook/sam-vit-base")
 
 To do point prompting, pass the input point to the processor, then take the processor output
 and pass it to the model for inference. To post-process the model output, pass the outputs and
-`original_sizes` and `reshaped_input_sizes` we take from the processor's initial output. We need to pass these
+`original_sizes` we take from the processor's initial output. We need to pass these
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
`original_sizes` are taken from the processor's initial output. We need to pass these
```

File: docs/source/en/tasks/mask_generation.md [hunk 9]
```diff
@@ -235,3 +227,326 @@ plt.show()
 <div class="flex justify-center">
      <img src="https://huggingface.co/datasets/huggingface/documentation-images/resolve/main/transformers/tasks/box_inference.png" alt="Visualized Inference"/>
 </div>
+
+## Fine-tuning for Mask Generation 
+
+We will fine-tune SAM2.1 on small part of MicroMat dataset for image matting. We need to install monai library to use the DICE loss, and trackio for logging the masks during training.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
We will fine-tune SAM2.1 on small part of MicroMat dataset for image matting. We need to install the [monai](https://github.com/Project-MONAI/MONAI) library to use DICE loss, and [trackio](https://huggingface.co/docs/trackio/index) for logging the masks during training.
```

File: docs/source/en/tasks/mask_generation.md [hunk 10]
```diff
@@ -235,3 +227,326 @@ plt.show()
...
+ 'prompt': '{"point": [[137, 1165, 1], [77, 1273, 0], [58, 1351, 0]], "bbox": [0, 701, 251, 1356]}',
+ 'image_id': '0034',
+ 'object_id': '34',
+ 'sample_idx': 1,
+ 'granularity': 'fine',
+ 'image_path': '/content/MicroMat-mini/img/0034.png',
+ 'mask_path': '/content/MicroMat-mini/mask/0034_34.png',
+ 'prompt_path': '/content/MicroMat-mini/prompt/0034_34.json'}
+```
+Prompts are string of dictionaries, so can get bounding boxes like below.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
Prompts are string of dictionaries, so you can get the bounding boxes as shown below.
```

File: docs/source/en/tasks/mask_generation.md [hunk 11]
```diff
@@ -235,3 +227,326 @@ plt.show()
...
+fig, ax = plt.subplots()
+ax.imshow(image)
+show_mask(ground_truth_mask, ax)
+ax.set_title("Ground truth mask")
+ax.set_axis_off()
+
+plt.show() 
+```
+
+Now we can define our dataset for loading the data. SAMDataset wraps an our dataset and formats each sample the way the SAM processor expects. So instead of raw images and masks, you get processed images, bounding boxes, and ground-truth masks ready for training.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
Now we can define our dataset for loading the data. SAMDataset wraps our dataset and formats each sample the way the SAM processor expects. So instead of raw images and masks, you get processed images, bounding boxes, and ground-truth masks ready for training.
```

File: docs/source/en/tasks/mask_generation.md [hunk 12]
```diff
@@ -235,3 +227,326 @@ plt.show()
...
+
+By default, processor resizes images, so on top of images and masks, it also returns original sizes. We also need to binarize the mask as it has values 0, 255.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
By default, processor resizes images, so on top of images and masks, it also returns original sizes. We also need to binarize the mask as it has values [0, 255].
```

File: docs/source/en/tasks/mask_generation.md [hunk 13]
```diff
@@ -235,3 +227,326 @@ plt.show()
...
+We can initialize the processor and the dataset with it. 
+
+```python 
+from transformers import Sam2Processor
+
+processor = Sam2Processor.from_pretrained("facebook/sam2.1-hiera-small")
+train_dataset = SAMDataset(dataset=train_ds, processor=processor)
+``` 
+
+We need to define a data collator that will turn varying size of ground truth masks to batches of reshaped masks in same shape. We reshape them using nearest neighbor interpolation. We also make batched tensors for rest of the elements in the batch. If your masks are all of same size, feel free to skip this step.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md:349]:
```suggestion
We need to define a data collator that will turn varying sizes of ground truth masks to batches of reshaped masks in same shape. We reshape them using nearest neighbor interpolation. We also make batched tensors for rest of the elements in the batch. If your masks are all the same size, feel free to skip this step.
```

File: docs/source/en/tasks/mask_generation.md [hunk 14]
```diff
@@ -235,3 +227,326 @@ plt.show()
...
+for k,v in batch.items():
+  print(k,v.shape)
+
+# pixel_values torch.Size([4, 3, 1024, 1024])
+# original_sizes torch.Size([4, 1, 2])
+# input_boxes torch.Size([4, 1, 4])
+# ground_truth_mask torch.Size([4, 1, 256, 256])
+#original_image_size torch.Size([4, 2])
+```
+We will now load the model, we will freeze the vision and the prompt encoder and only train the mask decoder.
```
Comment (stevhliu) [inline docs/source/en/tasks/mask_generation.md]:
```suggestion
We will now load the model and freeze the vision and the prompt encoder to only train the mask decoder. 
```

File: docs/source/en/tasks/video_text_to_text.md [hunk 15]
```diff
@@ -18,9 +18,13 @@ rendered properly in your Markdown viewer.
...
+These models have nearly the same architecture as [image-text-to-text](../image_text_to_text) models except for some changes to accept video data, since video data is essentially image frames with temporal dependencies. Some image-text-to-text models take in multiple images, but this alone is inadequate for a model to accept videos. 
+
+Moreover, video-text-to-text models are often trained with all vision modalities. Each example might have videos, multiple videos, images and multiple images. Some of these models can also take interleaved inputs. For example, you can refer to a specific video inside a string of text by adding a video token in text like "What is happening in this video? `<video>`".
+
+Note that these models process videos with no audio. [any-to-any](../any-to-any) models on the other hand can process videos with audio in them.
```
Comment (stevhliu) [inline docs/source/en/tasks/video_text_to_text.md]:
```suggestion
Note that these models process videos with no audio. [Any-to-any](../any-to-any) models on the other hand can process videos with audio in them.
```

File: docs/source/en/tasks/video_text_to_text.md [hunk 16]
```diff
@@ -120,28 +70,99 @@ Both videos have cats.
   </div>
 </div>
 
-Now we can preprocess the inputs.
 
-This model has a prompt template that looks like following. First, we'll put all the sampled frames into one list. Since we have eight frames in each video, we will insert 12 `<image>` tokens to our prompt. Add `assistant` at the end of the prompt to trigger the model to give answers. Then we can preprocess.
+Videos are series of image frames. Depending on the hardware limitations, downsampling is required. If the number of downsampled frames are too little, predictions will be low quality. 
+
+
+`video-text-to-text` models have processors with video processor abstracted in them. You can pass video inference related arguments to [`~ProcessorMixin.apply_chat_template`] function.
```
Comment (stevhliu) [inline docs/source/en/tasks/video_text_to_text.md]:
```suggestion
Video-text-to-text models have processors with video processor abstracted in them. You can pass video inference related arguments to [`~ProcessorMixin.apply_chat_template`] function.
```

File: docs/source/en/tasks/video_text_to_text.md [hunk 17]
```diff
@@ -120,28 +70,99 @@ Both videos have cats.
...
+
+<Tip warning={true}>
 
+You can learn more about video processors [here](../main_classes/video_processor).
+
+</Tip>
```
Comment (stevhliu) [inline docs/source/en/tasks/video_text_to_text.md]: We use this format in the docs now :)

```suggestion
> [!WARNING]
> You can learn more about video processors [here](../main_classes/video_processor).
```

File: docs/source/en/tasks/video_text_to_text.md [hunk 18]
```diff
@@ -120,28 +70,99 @@ Both videos have cats.
...
+
+We can define our chat template, passing in video with a URL like below.
```
Comment (stevhliu) [inline docs/source/en/tasks/video_text_to_text.md]:
```suggestion
We can define our chat history, passing in video with a URL like below.
```

**Review by stevhliu — APPROVED** [review]
Really nice, thanks for cooking! 🐈

File: docs/source/en/tasks/image_text_to_text.md (hunk 3 above)
Comment (merveenoyan) [reply docs/source/en/tasks/image_text_to_text.md]: you're right, I recently benchmarked myself, removing
